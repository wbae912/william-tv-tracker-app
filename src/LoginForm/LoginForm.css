.login-section {
  box-sizing: border-box;
  font-family: 'Montserrat', sans-serif;
  color: white;
  text-align: center;
  width: 90%;
  margin-right: auto;
  margin-left: auto;
  margin-top: 5%;
}

.login-h2 {
  margin-top: 0;
  margin-top: 4%
}

/* .login__user_name {
  margin-top: 2.5%;
  display: flex;
  justify-content: center;
}

.login__password {
  margin-top: 2.5%;
  display: flex;
  justify-content: center;
} */

.login-label {
  display: inline-block;
  width: 150px;
}

.login-input {
  margin-left: 2%;
  margin-right: 2%;
}

.buttons-flex {
  display: flex;
  margin-top: 2%;
}

.buttons-flex > .back-button-div, .login-button-div {
  flex: 1
}

.login-button-div {
  display: flex;
  justify-content: flex-end;
}

.back-button-div {
  display: flex;
  justify-content: flex-start;
}

.form__login-button, .back-button {
  background-color: transparent;
  color: white;
  border-radius: 10px;
  border: 1.5px solid white;
  width: 80px;
  height: 30px;
  font-size: 14px;
  transition-duration: 0.4s;
}

.form__login-button {
  margin-right: 1%;
}

.back-button {
  margin-left: 1%;
}

.form__login-button:hover, .back-button:hover {
  background-color: white;
  color: black;
}