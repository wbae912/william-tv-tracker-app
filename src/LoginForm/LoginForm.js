import React, { Component } from 'react';
import TokenService from '../services/token-service';
import AuthApiService from '../services/auth-api-service';
import tvContext from '../Context';
import FormError from '../FormError/FormError';
import './LoginForm.css';

export default class LoginForm extends Component {
  static contextType = tvContext;

  constructor(props) {
    super(props)
  
    this.state = {
       error: null,
       password: '',
       user_name: '',
       usernameTouch: {
        touched: false
       },
       passwordTouch: {
         touched: false
       }
    }
  }
  
  handleInputChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  handleUserNameTouch = (e) => {
    this.setState({
      usernameTouch: {touched: true}
    })
  }

  handlePasswordTouch = (e) => {
    this.setState({
      passwordTouch: {touched: true}
    })
  }
  
  handleSubmitJwtAuth = ev => {
    ev.preventDefault()
    const { user_name, password } = ev.target

    AuthApiService.postLogin({
      user_name: user_name.value,
      password: password.value,
    })
      .then(res => {
        user_name.value = '';
        password.value = '';
        TokenService.saveAuthToken(res.authToken);
        this.props.history.push('/dashboard')
      })
      .catch(res => {
        this.setState({
          error: res.error
        })
      })
    }

  validateUserName = () => {
    const username = this.state.user_name;
    if(username.length === 0) {
      return 'Please enter a username';
    }
  }

  validatePassword = () => {
    const password = this.state.password;
    const REGEX = /(?=.*\d)(?=.*[a-z])(?=.*[A-Z])/
    if(password.length === 0) {
        return 'Please enter a password';
    } else if(password.length < 8 || password.length > 72) {
        return 'Password must be between 8 and 72 characters';
    }
    if (password.startsWith(' ') || password.endsWith(' ')) {
      return 'Password must not start or end with empty spaces';
    }
    if (!REGEX.test(password)) {
      return 'Password must contain at least one upper case, lower case, and number character';
    }
  }


  render() {
    let passwordError = this.validatePassword();
    let userNameError = this.validateUserName();

    return (
      <section className="login-section">
        <div role="alert" className="error-div">
            {this.state.error && <p className='red'>{this.state.error}</p>}
          </div>
        <h2 className="login-h2">Login</h2>
        <form className="login-form" onSubmit={(e) => this.handleSubmitJwtAuth(e)}>
          <div className="login__user_name">
            <label htmlFor="user_name" className="login-label">Username</label>
            <input type="text" name="user_name" id="user_name" className="login-input" required
              onChange={(e) => {this.handleInputChange(e); this.handleUserNameTouch(e)}}
            />
            {this.state.usernameTouch.touched && <FormError message={userNameError} />}
          </div>
          <div className="login__password">
            <label htmlFor="password" className="login-label">Password</label>
            <input type="password" name="password" id="password" minLength="8" maxLength="72" className="login-input" 
              autoComplete="on" required 
              onChange={(e) => {this.handleInputChange(e); this.handlePasswordTouch(e)}}
            />
            {this.state.passwordTouch.touched && <FormError message={passwordError} />}
          </div>
          <div className="buttons-flex">
            <div className="login-button-div">
              <button type="submit" className="form__login-button" disabled={userNameError || passwordError}>Login</button>
            </div>
            <div className="back-button-div">
              <button type="button" className="back-button" onClick={() => this.props.history.push('/')}>Back</button>
            </div>
          </div>
        </form>
      </section>
    )
  }
}