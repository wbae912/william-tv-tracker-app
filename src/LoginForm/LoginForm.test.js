import React from 'react';
import ReactDOM from 'react-dom';
import LoginForm from './LoginForm';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';

describe('LoginForm component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<MemoryRouter><LoginForm /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  
  it('renders UI as expected', () => {
    const tree = renderer
      .create(<MemoryRouter><LoginForm /></MemoryRouter>)
      .toJSON();
      expect(tree).toMatchSnapshot();
  })
})