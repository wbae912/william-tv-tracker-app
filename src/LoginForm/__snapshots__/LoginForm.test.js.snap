// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component renders UI as expected 1`] = `
<section
  className="login-section"
>
  <div
    className="error-div"
    role="alert"
  />
  <h2
    className="login-h2"
  >
    Login
  </h2>
  <form
    className="login-form"
    onSubmit={[Function]}
  >
    <div
      className="login__user_name"
    >
      <label
        className="login-label"
        htmlFor="user_name"
      >
        Username
      </label>
      <input
        aria-invalid="true"
        aria-label="username"
        aria-required="true"
        className="login-input"
        id="user_name"
        name="user_name"
        onChange={[Function]}
        required={true}
        type="text"
      />
    </div>
    <div
      className="login__password"
    >
      <label
        className="login-label"
        htmlFor="password"
      >
        Password
      </label>
      <input
        aria-invalid="true"
        aria-label="password"
        aria-required="true"
        autoComplete="on"
        className="login-input"
        id="password"
        maxLength="72"
        minLength="8"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
      />
    </div>
    <div
      className="buttons-flex"
    >
      <div
        className="login-button-div"
      >
        <button
          className="form__login-button"
          disabled="Please enter a username"
          type="submit"
        >
          Login
        </button>
      </div>
      <div
        className="back-button-div"
      >
        <button
          className="back-button"
          onClick={[Function]}
          type="button"
        >
          Back
        </button>
      </div>
    </div>
  </form>
</section>
`;
